### PR TITLE
Backport upstream improvements to pg_dump

### PR DESCRIPTION
**Pull request by bmdoil**
The purpose of this PR is to backport a series of refactors and performance improvements made to pg_dump from upstream REL_12_STABLE through REL_15_BETA1.

Before:
```
t/001_basic.pl ......... ok
t/002_pg_dump.pl ....... ok
t/010_dump_connstr.pl .. ok
All tests successful.
Files=3, Tests=6093, 41 wallclock secs ( 0.19 usr  0.02 sys +  5.46 cusr  0.99 csys =  6.66 CPU)
Result: PASS
pg_dump -s regression > /dev/null  0.59s user 0.12s system 2% cpu 33.560 total
```

After:
```
t/001_basic.pl ......... ok
t/002_pg_dump.pl ....... ok
t/010_dump_connstr.pl .. ok
All tests successful.
Files=3, Tests=6248, 27 wallclock secs ( 0.19 usr  0.02 sys +  5.47 cusr  0.95 csys =  6.63 CPU)
Result: PASS
pg_dump -s regression > /dev/null  0.44s user 0.07s system 6% cpu 8.440 total
```

Below is a list of the backported upstream commits. Please refer to individual commit messages for details and resolved conflicts.

https://github.com/postgres/postgres/commit/d5e8930 pg_dump: minor performance improvements from eliminating sub-SELECTs.
https://github.com/postgres/postgres/commit/e2c52be pg_dump: Refactor getIndexes()
https://github.com/postgres/postgres/commit/be85727 Use PREPARE/EXECUTE for repetitive per-object queries in pg_dump.
https://github.com/postgres/postgres/commit/9895961 Avoid per-object queries in performance-critical paths in pg_dump.
https://github.com/postgres/postgres/commit/0c9d844 Rethink pg_dump's handling of object ACLs.
https://github.com/postgres/postgres/commit/5209c0b Refactor pg_dump's tracking of object components to be dumped.
https://github.com/postgres/postgres/commit/a22d6a2 pg_dump: Add missing relkind case
https://github.com/postgres/postgres/commit/70bef49 Fix minor memory leaks in pg_dump.
https://github.com/postgres/postgres/commit/92316a4 In pg_dump, use simplehash.h to look up dumpable objects by OID.
https://github.com/postgres/postgres/commit/2acc84c pg_dump: fix mis-dumping of non-global default privileges.
https://github.com/postgres/postgres/commit/4438eb4 pg_dump: Reorganize getTables()
https://github.com/postgres/postgres/commit/6b96673 In pg_dump, avoid doing per-table queries for RLS policies.
https://github.com/postgres/postgres/commit/6c450a8 Cache the results of format_type() queries in pg_dump.
https://github.com/postgres/postgres/commit/d782d59 Avoid invoking PQfnumber in loop constructs
https://github.com/postgres/postgres/commit/6499008 pg_dump: Add const decorations
https://github.com/postgres/postgres/commit/0bf8364 pg_dump: Fix dumping of inherited generated columns
https://github.com/postgres/postgres/commit/8e396a7 pg_dump: label PUBLICATION TABLE ArchiveEntries with an owner.
https://github.com/postgres/postgres/commit/9eabfe300a22ad3d776dc293265e15379790bd9a pg_dump: label INDEX ATTACH ArchiveEntries with an owner.
https://github.com/postgres/postgres/commit/d5ab79d815783fe60062cefc423b54e82fbb92ff Make pg_dump's table of object-type priorities more maintainable.
https://github.com/postgres/postgres/commit/04732962462ba99cf8f8bcf6ac83932867cc96a8 pg_dump: Reorganize dumpBaseType()
https://github.com/postgres/postgres/commit/8354e7b27ebec8bf78088e22b5b9c1caa46db8e1 Remove unused parameters
https://github.com/postgres/postgres/commit/5423853feebd30772b7ff9b306885dcb02b79e76 Avoid retrieval of CHECK constraints and DEFAULT exprs in data-only dump
https://github.com/postgres/postgres/commit/ed2c7f65bd9f15f8f7cd21ad61602f983b1e72e9 pg_dump: Reorganize dumpFunc() and dumpAgg()
https://github.com/postgres/postgres/commit/5fea14f4b2c5ca4e4d9def92b17e50fcfbaf3468 Avoid trying to restore table ACLs and per-column ACLs in parallel.
https://github.com/postgres/postgres/commit/991c444e7ad3a1ff7bcffff4121e6ff67c4783fc pg_dump: Further reorganize getTableAttrs()
https://github.com/postgres/postgres/commit/d9fa17aa7c34dea66ce64da6fb4c643e75ba452c pg_dump: Remove dead code
https://github.com/postgres/postgres/commit/b08dee24a5574ba6657d238b91c10f1f8e2608f8 Add pg_dump support for ALTER obj DEPENDS ON EXTENSION
https://github.com/postgres/postgres/commit/f31364676dfd9e7a2b712ea0da7b4ab6b7d28f5e Teach pg_dump to dump comments on RLS policy objects.
https://github.com/postgres/postgres/commit/8f4ee44bcdbcd437f860f2d048125a118335b088 Fix minor memory leak in pg_dump
https://github.com/postgres/postgres/commit/07b39083c2aca003c4b1f289d7dc2368b5e2297a Fix pg_dump's handling of dependencies for custom opclasses.
 postgres/postgres@9a4c0e3 Dump ALTER TABLE ... ATTACH PARTITION as a separate ArchiveEntry.
 postgres/postgres@d5ab79d Make pg_dump's table of object-type priorities more maintainable.
 postgres/postgres@9e5f1f2 Rethink recent fix for pg_dump's handling of extension config tables.
 postgres/postgres@3eb3d3e Collect attribute data on extension owned tables being dumped
 postgres/postgres@3dadcb2 pg_dump: fix dependencies on FKs to partitioned tables
 postgres/postgres@b304b2b Fix parallel restore of FKs to partitioned tables
 
In addition to the backports:
* Pre-GPDB5 version specific code has been removed.
* The GPDB specific TypeStorageOptions struct and supporting code has been removed in favor of adding
  pg_type_encoding.typoptions to the getTypes query. The usage of CatalogId as the hash index caused collisions
  with the TypeInfo indexes when TypeStorageOptions was a unique dumpable object, since the only unique aspect was the
  typstorage field. Now, when dumpType is called, we check the typstorage field and dump the ALTER TABLE...SET DEFAULT 
  ENCODING TOC entry if needed.
* The following unused flags have been removed:
  - pre-data-schema-only
  - post-data-schema-only

### CODE REVIEW

File: src/bin/pg_dump/pg_dump.c [hunk 1]
```diff
@@ -15123,7 +15116,6 @@ dumpAgg(Archive *fout, AggInfo *agginfo)
 	aggmtransspace = PQgetvalue(res, 0, i_aggmtransspace);
 	agginitval = PQgetvalue(res, 0, i_agginitval);
 	aggminitval = PQgetvalue(res, 0, i_aggminitval);
-	convertok = (PQgetvalue(res, 0, i_convertok)[0] == 't');
 
 	if (fout->remoteVersion >= 80400)
```
Comment (soumyadeep2007) [inline src/bin/pg_dump/pg_dump.c:14500]: > Remove dumpAgg query block for 80100 server version

Did you remove the block in some other commit? I don't see the hunk here.

File: src/bin/pg_dump/pg_dump.c [hunk 2]
```diff
@@ -12766,173 +12666,81 @@ dumpFunc(Archive *fout, FuncInfo *finfo)
 
 	/* GPDB_95_MERGE_FIXME: use isGE70 instead? */
```
Comment (soumyadeep2007) [inline src/bin/pg_dump/pg_dump.c]: Can this FIXME be removed now?

File: src/bin/pg_dump/pg_dump.c [hunk 3]
```diff
@@ -14926,152 +14655,104 @@ dumpAgg(Archive *fout, AggInfo *agginfo)
 	details = createPQExpBuffer();
 
 	/* Get aggregate-specific details */
```
Comment (soumyadeep2007) [inline src/bin/pg_dump/pg_dump.c]: This is a merge hazard zone (if certain agg fields were backported they might not appear under the correct version guard) - not saying that there's anything wrong. Just flagging for closer inspection.

**Review by soumyadeep2007 — COMMENTED** [review]
I will keep updating this high-level comment each day. It can be edited by
anyone and inline replies can be posted here (to prevent a comment explosion).
Just add: "[your_name]" at the beginning to reply inline. After a discussion,
if I feel that a topic is resolved, I will put a [✓] next to it.

**I think the overall intent of this PR is to improve pg_dump w/ pg_upgrade
primarily in mind.** So there might be some backports that improve pg_dump or
pg_restore outside that context - probably these can be left alone? (unless it
creates more conflicts and more work). Thoughts?

Rambling (there maybe a better way):

An easy way to verify perhaps if a commit can be removed from the chain is to
just drop it during a rebase and put a break before the next commit in the
chain - and then try running compile and tests at that point. Furthermore,
putting a break before the last commit and rerunning tests/compile? If there is
a conflict at any stage, abort the rebase.

**Maybe certain backports could be avoided for master and 6X (for now):**

* Avoid trying to restore table ACLs and per-column ACLs in parallel. - Only
applies to fix a bug w/ parallel pg_restore, which is not invoked by pg_upgrade.

**These don't make sense for 6X, but okay for master:**

* Fix parallel restore of FKs to partitioned tables - FKs weren't supported on
  partitioned tables until PG12.

* pg_dump: fix dependencies on FKs to partitioned tables - FKs weren't supported
  on partitioned tables until PG12.

* Teach pg_dump to dump comments on RLS policy objects. - RLS isn't a 6X thing

* Add pg_dump support for ALTER obj DEPENDS ON EXTENSION - Not a 6X capability

* Collect attribute data on extension owned tables being dumped - Bugfix for a
12 introduced bug

**Reviewed up to `Avoid retrieval of CHECK constraints and DEFAULT exprs in data-only dump`**

File: src/bin/pg_dump/pg_dump.c (hunk 1 above)
Comment (bmdoil) [reply src/bin/pg_dump/pg_dump.c:14500]: It's been removed in `Remove unnecessary version checks from pg_dump`

**Review by soumyadeep2007 — COMMENTED** [review]
Reviewed upto `Refactor pg_dump's tracking of object components to be dumped.`

**Review by soumyadeep2007 — COMMENTED** [review]
> * Rejected changes to special handling of public
  schema from upstream a7a7be1f2 and b073c3ccd
  which have not been backported.

Can you add an explanation as to why to the commit?
 `Rethink pg_dump's handling of object ACLs.`
[Brent] Added more details and linked to the dev thread.

File: src/bin/pg_dump/pg_dump.c [hunk 4]
```diff
@@ -6346,72 +6207,40 @@ getExtProtocols(Archive *fout, int *numExtProtocols)
...
+								 "oid, "
+								 "ptcname, "
+								 "ptcreadfn as ptcreadoid, "
+								 "ptcwritefn as ptcwriteoid, "
+								 "ptcvalidatorfn as ptcvaloid, "
+								 "(%s ptcowner) as rolname, "
+								 "ptctrusted, ptcacl, "
+								 "acldefault('E', ptcowner) AS acldefault "
+								 "FROM pg_extprotocol ", 
+								 username_subquery);
```
[the remark this answers is not in the document]
Comment (bmdoil) [reply src/bin/pg_dump/pg_dump.c]: Correct, and will verify with a test.

File: src/bin/pg_dump/pg_dump.c [hunk 5]
```diff
@@ -5704,36 +5705,39 @@ getTypeStorageOptions(Archive *fout, int *numTypes)
...
+					  "ELSE pg_catalog.format_type(t.oid, NULL) "
+					  "END  as typname, "
+					  "t.tableoid as tableoid, "
+					  "t.oid AS oid, "
+					  "t.typnamespace AS typnamespace, "
+					  "(%s typowner) as rolname, "
+					  "array_to_string(a.typoptions, ', ') AS typoptions "
+					  "FROM pg_type t "
+					  "JOIN pg_catalog.pg_type_encoding a ON a.typid = t.oid "
+					  "WHERE t.typisdefined = 't'", username_subquery);
```
[the remark this answers is not in the document]
Comment (bmdoil) [reply src/bin/pg_dump/pg_dump.c]: Added `ALTER TYPE dump_test.int42 SET DEFAULT ENCODING` test to `002_pg_dump.pl`

File: src/bin/pg_dump/pg_dump.c [hunk 6]
```diff
@@ -10954,6 +10872,11 @@ dumpType(Archive *fout, const TypeInfo *tyinfo)
 	else
 		pg_log_warning("typtype of data type \"%s\" appears to be invalid",
 					   tyinfo->dobj.name);
+
+	if (tyinfo->typstorage && *tyinfo->typstorage != '\0')
+	{
```
[the remark this answers is not in the document]
Comment (bmdoil) [reply src/bin/pg_dump/pg_dump.c:10809]: Fixed

**Review by soumyadeep2007 — COMMENTED** [review]
1st pass done. Reviewed all commits. Will give approval once all comments are addressed and I can respond to your replies.

### `Rethink pg_dump's handling of object ACLs.`

Can you add an explanation as to why we are rejecting these in the commit?
 

* Rejected changes to getTriggers introduced by f0e21f2f6
  which has not been backported.

* Handle GPDB specific attencoding field in getTableAttrs.
[Brent]: Updated the `Avoid per-object queries in performance-critical paths in pg_dump.` commit message.
New message:
```
    Conflicts resolved:
    * Handle GPDB specific attencoding field in getTableAttrs.
    * Rejected changes to getTriggers introduced by f0e21f2f6
      which has not been backported because triggers are not
      supported in GPDB.
    * Rejected attcompression field in getTableAttrs which
      was introduced in PG14 and has not been backported because
      it is not supported in GPDB.
```

>  It might seem that WHERE clauses listing many target
OIDs could be inefficient, but at least on recent server versions
this provides a very substantial speedup.

Do we have these changes from these "recent server versions"? If not, we might
need to cherry-pick them too, if feasible.
[Brent]: I looked at this a bit, "recent server versions" wasn't super helpful so I'm going to leave it to the 6X backport and the test framework we already have in place to determine the performance. I can't think of a situation where it will be worse than what we have now...

> The new WHERE clauses depend on the unnest() function, which is
only present in 8.4 and above.  We could implement them differently
for older servers, but there is an ongoing discussion that will
probably result in dropping pg_dump support for servers before 9.2,
so that seems like it'd be wasted work.  For now, just bump the
server version check to require >= 8.4, without stopping to remove
any of the code that's thereby rendered dead.  We'll mop that
situation up soon.

We get lucky here. 5X actually has unnest().
[Brent]: Yes! Was very happy to see that.

```
postgres=# SELECT unnest(ARRAY[1,2]);
 unnest 
--------
      1
      2
(2 rows)
```

Did you address this aspect? I couldn't find the min version bump - do we need
to lower it to 8.3. If we can relax this to apply the unnest goodness to 5X ->
easier backport and more performant queries! Yay! Also adding a comment in the
code and a comment to the conflicts section would be really useful.

[Brent] - Yes, I checked unnest for GPDB5 early on and lowered the min version to 8.3. With that said I haven't fully tested a binary dump from GPDB5. There will likely be changes needed to enable it because of general differences in master. I hit `s/gp_role/gp_session_role` immediately. 

### `Use PREPARE/EXECUTE for repetitive per-object queries in pg_dump.`

> Reject changes to dumpTableAttach, which
  added a query for partbound details in
  e3fcbbd623b9ccc16cdbda374654d91a4727d173

 Lets mention that e3fcbbd623b9ccc16cdbda374654d91a4727d173 wasn't backported
 as it was considered too invasive, as the reason for the rejection.
[Brent]: Done

 > This patch applies the idea to functions, aggregates, operators, and
data types.  While it could be carried further, the remaining sorts of
objects aren't likely to appear in typical databases enough times to
be worth worrying over.  Moreover, doing the PREPARE is likely to be a
net loss if there aren't at least some dozens of objects to apply the
prepared query to.

Are we going to apply this mechanism to GPDB specific things in follow-up
commits? Also what are the GPDB specific things that would benefit from this? If
you have already identified those, it would be nice to document that in the
commit message.
[Brent]: We can apply this to GPDB specific things where it makes sense, but I haven't identified yet what those are. The follow-up PR to improve binary upgrade calls can address that also.

File: src/bin/pg_dump/pg_dump.c [hunk 7]
```diff
@@ -6346,72 +6207,40 @@ getExtProtocols(Archive *fout, int *numExtProtocols)
 	int			i_ptcname;
 	int			i_rolname;
 	int			i_ptcacl;
-	int			i_ptcracl;
-	int			i_ptcinitacl;
-	int			i_ptcinitracl;
+	int			i_acldefault;
```
[the remark this answers is not in the document]
Comment (bmdoil) [reply src/bin/pg_dump/pg_dump.c:6179]: Added a dump test for protocol and external web table

File: src/bin/pg_dump/pg_dump.c [hunk 8]
```diff
@@ -331,23 +327,27 @@ static void binary_upgrade_set_pg_class_oids(Archive *fout,
...
 
 
 /* START MPP ADDITION */
+static void dumpExtProtocol(Archive *fout, ExtProtInfo *ptcinfo);
+static void dumpTypeStorageOptions(Archive *fout, TypeStorageOptions *tstorageoptions);
+static void dumpPreassignedOidArchiveEntry(Archive *fout, BinaryUpgradeInfo *binfo);
+static void dumpPreassignedOidDefinition(Archive *fout, BinaryUpgradeInfo *binfo);
 static void setExtPartDependency(TableInfo *tblinfo, int numTables);
-static char *format_table_function_columns(Archive *fout, FuncInfo *finfo, int nallargs,
+static char *format_table_function_columns(Archive *fout, const FuncInfo *finfo, int nallargs,
```
[the remark this answers is not in the document]
Comment (soumyadeep2007) [reply src/bin/pg_dump/pg_dump.c:366]: Lets track this internally.

**Review by soumyadeep2007 — APPROVED** [review]
Can you please add the assertion failure you were getting as additional context in the commit message for the gp_role thing?

Thanks for all the hard work on this! LGTM! :tada:

**Pull request merged**